botocore/docs/bcdoc/docstringparser.py

```diff
@@ -115,8 +115,14 @@ def write(self, doc):
         self._write_children(doc)
 
     def _write_children(self, doc):
-        for child in self.children:
-            child.write(doc)
+        for index in range(len(self.children)):
+            if isinstance(self.children[index], TagNode):
+                next_child = None
+                if index + 1 < len(self.children):
+                    next_child = self.children[index + 1]
+                self.children[index].write(doc, next_child)
+            else:
+                self.children[index].write(doc)
```

Comment:
We might be able to simplify this logic a little, since we're already accessing the value at the target every time, we can use enumerate to grab it for us.

```suggestion
        for index, child in enumerate(self.children):
            next_child = None
            if isinstance(child, TagNode) and index + 1 < len(self.children):
                next_child = self.children[index + 1]
            child.write(doc, next_child)
```

Reply:
nit: set `index + 1` to a variable since we're using it multiple times. Maybe `next_child_index`?

Reply:
This simplified logic would require adding an optional parameter to `DataNode.write(self, doc, next_child=None)` which would never be used since it currently only accepts `(self, doc)`. Is this a good tradeoff for a simplified logic?

Reply:
Yeah, if we're not overriding every function signature, let's reintroduce the if/else and call without `next_child` if it's not a TagNode.

Reply:
> nit: set index + 1 to a variable since we're using it multiple times. Maybe next_child_index?

Generally, we create variables for complex concepts like multi-part conditionals. That let's us easily understand what it's trying to express. For things like incrementing a counter, there's minimal performance hit in doing it repeatedly (in some cases it may take more time to allocate the variable than the temporary register value).

It's likely easier to read this way, so unless we have a compelling argument, it's probably fine to leave.